extensions/void/src/common/sendLLMMessage.ts

```diff
@@ -220,66 +219,85 @@ const sendGreptileMsg: SendLLMMessageFnTypeInternal = ({ messages, onText, onFin
 export const sendLLMMessage: SendLLMMessageFnTypeExternal = ({ messages, onText, onFinalMessage, apiConfig }) => {
 	if (!apiConfig) return { abort: () => { } }
 
-	const whichApi = apiConfig.whichApi
-
-	if (whichApi === 'anthropic') {
-		return sendClaudeMsg({ messages, onText, onFinalMessage, apiConfig })
-	}
-	else if (whichApi === 'openai') {
-		return sendOpenAIMsg({ messages, onText, onFinalMessage, apiConfig })
-	}
-	else if (whichApi === 'greptile') {
-		return sendGreptileMsg({ messages, onText, onFinalMessage, apiConfig })
+	const whichApi = apiConfig.whichApi;
+
+	switch (whichApi) {
+		case 'anthropic':
+			return sendClaudeMsg({ messages, onText, onFinalMessage, apiConfig });
+		case 'openai':
+			return sendOpenAIMsg({ messages, onText, onFinalMessage, apiConfig });
+		case 'greptile':
+			return sendGreptileMsg({ messages, onText, onFinalMessage, apiConfig });
+		case 'ollama':
+			return sendOllamaMsg({ messages, onText, onFinalMessage, apiConfig });
+		default:
+			console.error(`Error: whichApi was ${whichApi}, which is not recognized!`);
+			return sendClaudeMsg({ messages, onText, onFinalMessage, apiConfig }); // TODO
 	}
-	else if (whichApi === 'ollama') {
-		return sendClaudeMsg({ messages, onText, onFinalMessage, apiConfig }) // TODO
-	}
-	else {
-		console.error(`Error: whichApi was ${whichApi}, which is not recognized!`)
-		return sendClaudeMsg({ messages, onText, onFinalMessage, apiConfig }) // TODO
-	}
-
 }
 
 
 // Ollama
-// const sendOllamaMsg: sendMsgFnType = ({ messages, onText, onFinalMessage }) => {
-
-//     let did_abort = false
-//     let fullText = ''
-
-//     // if abort is called, onFinalMessage is NOT called, and no later onTexts are called either
-//     let abort: () => void = () => {
-//         did_abort = true
-//     }
-
-//     ollama.chat({ model: 'llama3.1', messages: messages, stream: true })
-//         .then(async response => {
-
-//             abort = () => {
-//                 // response.abort() // this isn't needed now, to keep consistency with claude will leave it commented for now
-//                 did_abort = true;
-//             }
-
-//             // when receive text
-//             try {
-//                 for await (const part of response) {
-//                     if (did_abort) return
-//                     let newText = part.message.content
-//                     fullText += newText
-//                     onText(newText, fullText)
-//                 }
-//             }
-//             // when error/fail
-//             catch (e) {
-//                 onFinalMessage(fullText)
-//                 return
-//             }
-
-//             // when we get the final message on this stream
-//             onFinalMessage(fullText)
-//         })
-
-//     return { abort };
-// };
+export const sendOllamaMsg: SendLLMMessageFnTypeInternal = ({ messages, onText, onFinalMessage, apiConfig }) => {
```

Comment:
Could be simpler to use `ollama-js` here for future maintainability:
```javascript
import ollama from 'ollama/browser';

const sendOllamaMsg: SendLLMMessageFnTypeInternal = ({ messages, onText, onFinalMessage, apiConfig }) => {
	const ollama = new Ollama({ host: apiConfig.ollama.host });

	let did_abort = false
	let fullText = ''

	// if abort is called, onFinalMessage is NOT called, and no later onTexts are called either
	let abort: () => void = () => { did_abort = true }

	ollama.chat({ model: apiConfig.ollama.model, messages: messages, stream: true })
		.then(async response => {

			abort = () => {
				ollama.abort();
				did_abort = true;
			}

			try {
				for await (const part of response) {
					if (did_abort) return
					let newText = part.message.content
					fullText += newText
					onText(newText, fullText)
				}
			}
			// when error/fail
			catch (e) {
				onFinalMessage(fullText)
				return
			}

			// when we get the final message on this stream
			onFinalMessage(fullText)
		})

	return { abort };
}
```

Reply:
Ahh good call. I was originally doing it this way but importing from 'ollama' and it was causing issues. Ill update this now. Thanks!

Reply:
1. @BruceMacD Great call. @w1gs, we should definitely use `ollama/browser` instead of calling `fetch()`. This should be easy to swap out using Bruce's  code. My only change would be to make it use `.try().catch()` to go with the convention of the rest of the file.

> 
> ```js
> import ollama from 'ollama/browser';
> 
> const sendOllamaMsg: SendLLMMessageFnTypeInternal = ({ messages, onText, onFinalMessage, apiConfig }) => {
> 	const ollama = new Ollama({ host: apiConfig.ollama.host });
> 
> 	let did_abort = false
> 	let fullText = ''
> 
> 	// if abort is called, onFinalMessage is NOT called, and no later onTexts are called either
> 	let abort: () => void = () => { did_abort = true }
> 
> 	ollama.chat({ model: apiConfig.ollama.model, messages: messages, stream: true })
> 		.then(async response => {
> 
> 			abort = () => {
> 				ollama.abort();
> 				did_abort = true;
> 			}
> 
> 			try {
> 				for await (const part of response) {
> 					if (did_abort) return
> 					let newText = part.message.content
> 					fullText += newText
> 					onText(newText, fullText)
> 				}
> 			}
> 			// when error/fail
> 			catch (e) {
> 				onFinalMessage(fullText)
> 				return
> 			}
> 
> 			// when we get the final message on this stream
> 			onFinalMessage(fullText)
> 		})
> 
> 	return { abort };
> }
> ```



I'm not sure if there's a way to specify the Ollama endpoint in `ollama/browser`. If there's not I'd be fine getting rid of the `apiConfig.ollama.model` variable and assuming users are hosting Ollama on its default url.